src/execution.jl

```diff
@@ -41,45 +41,56 @@ isghosttype(dt) = !dt.mutable && sizeof(dt) == 0
-context. The `gridDim` and `blockDim` arguments represent the launch configuration, the
-optional `shmem` parameter specifies how much bytes of dynamic shared memory should be
-allocated (defaulting to 0), while the optional `stream` parameter indicates on which stream
-the launch should be scheduled.
+context. The `@cuda` macro should prefix a kernel invocation, with one of the following
+arguments in the `kwargs` position:
+
+From `CUDAdrv.cudacall`:
+- threads (defaults to 1)
+- blocks (defaults to 1)
+- shmem (defaults to 0)
+- stream (defaults to the default stream)
+
+From `CUDAnative.cufunction`:
+- minthreads, maxtreads: specify the expected minimum resp. maximum amount of threads that
+  this kernel will be launched with. unspecified dimensions are assumed to be 1.
 
 The `func` argument should be a valid Julia function. It will be compiled to a CUDA function
 upon first use, and to a certain extent arguments will be converted and managed
 automatically (see [`cudaconvert`](@ref)). Finally, a call to `cudacall` is performed,
 scheduling the compiled function for execution on the GPU.
 """
-macro cuda(config::Expr, callexpr::Expr)
-    # sanity checks
-    if config.head != :tuple || !(2 <= length(config.args) <= 4)
-        throw(ArgumentError("first argument to @cuda should be a tuple (gridDim, blockDim, [shmem], [stream])"))
+macro cuda(ex...)
+    if length(ex) > 0 && ex[1].head == :tuple
+        error("The tuple argument to @cuda has been replaced by keywords: `@cuda threads=... fun(args...)`")
     end
-    if callexpr.head != :call
+    call = ex[end]
+    kwargs = ex[1:end-1]
+    if call.head != :call
         throw(ArgumentError("second argument to @cuda should be a function call"))
     end
 
-    # handle optional arguments and forward the call
-    # NOTE: we duplicate the CUDAdrv's default values of these arguments,
-    #       because the kwarg version of `cudacall` is too slow
-    stream = length(config.args)==4 ? esc(pop!(config.args)) : :(CuDefaultStream())
-    shmem  = length(config.args)==3 ? esc(pop!(config.args)) : :(0)
-    dims = esc(config)
-    args = :(cudaconvert.(($(map(esc, callexpr.args)...),)))
-    return :(_cuda($dims, $shmem, $stream, $args...))
+    return :(_cuda(cudaconvert.(($(map(esc, call.args)...),))...;
+                   $(map(esc, kwargs)...)))
 end
 
 const agecache = Dict{UInt, UInt}()
 const compilecache = Dict{UInt, CuFunction}()
-@generated function _cuda(dims::Tuple{CuDim, CuDim}, shmem, stream,
-                          func::Core.Function, argspec...)
+@generated function _cuda(func::Core.Function, argspec...; kwargs...)
     arg_exprs = [:( argspec[$i] ) for i in 1:length(argspec)]
     arg_types = argspec
 
+    # split kwargs, only some are dealt with by the compiler
+    nt = kwargs.parameters[4]
+    kws = Base._nt_names(nt)::NTuple{N,Symbol} where {N}
+    compile_kws = Symbol[kws...] ∩ (:minthreads,:maxthreads)    # JuliaLang/julia#25801
```

Comment:
There is also `minctasm` which is the second argument to `__launch_bounds__`. So far as I can tell nobody set `reqntid*`, at least Clang just leaves it be.

Reply:
Added. I don't think I'm going to support something like `__launch_bounds__` though, `minthreads` and `blocks_per_sm` etc look more user-friendly to me.